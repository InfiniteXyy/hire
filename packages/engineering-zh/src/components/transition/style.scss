@import '../../theme';

.fade-slide-enter {
  opacity: 0;
  transform: translateY(0.5rem);

  &--active {
    opacity: 1;
    transform: translateY(0);
    transition: opacity $quick-transition, transform $quick-transition;
  }
}

.fade-slide-exit {
  opacity: 1;
  transform: translateY(0);

  &--active {
    opacity: 0;
    transform: translateY(0.5rem);
    transition: opacity $quick-transition, transform $quick-transition;
  }
}
