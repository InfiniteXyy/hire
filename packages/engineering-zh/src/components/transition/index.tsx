import React, { useLayoutEffect, useRef, useState } from 'react';
import cls from 'classnames';
import './style.scss';

interface ContainerProps {
  visible: boolean;
  children: React.ReactNode;
}

export interface TransitionProps extends ContainerProps {
  cssPrefix: string;
}

type AnimStatus = 'None' | 'Begin' | 'MoveIn' | 'Finished' | 'MoveOut';
const DEFAULT_TRANSITION_DURATION = 150;

const mapStateToClass = (cssPrefix: string, status: AnimStatus): string => {
  switch (status) {
    case 'Begin':
      return `${cssPrefix}-enter`;
    case 'MoveIn':
      return `${cssPrefix}-enter--active`;
    case 'Finished':
      return `${cssPrefix}-exit`;
    case 'MoveOut':
      return `${cssPrefix}-exit--active`;
    default:
      return '';
  }
};

export function Transition(props: TransitionProps): JSX.Element {
  const { visible, children, cssPrefix } = props;
  const completeTimer = useRef<number>();
  const [animStatus, setAnimStatus] = useState<AnimStatus>('None');

  useLayoutEffect(() => {
    if (animStatus === 'Begin') {
      setTimeout(() => setAnimStatus('MoveIn'), 0);
    }
  }, [animStatus]);

  useLayoutEffect(() => {
    setAnimStatus(visible ? 'Begin' : 'MoveOut');
    clearTimeout(completeTimer.current);
    completeTimer.current = window.setTimeout(() => {
      setAnimStatus(visible ? 'Finished' : 'None');
    }, DEFAULT_TRANSITION_DURATION);
    return () => clearTimeout(completeTimer.current);
  }, [visible]);

  return (
    <>
      {animStatus !== 'None' &&
        React.cloneElement(children as React.ReactElement, {
          className: cls(
            (children as React.ReactElement).props.className || '',
            mapStateToClass(cssPrefix, animStatus)
          ),
        })}
    </>
  );
}

export function Fade(props: Omit<TransitionProps, 'cssPrefix'>): JSX.Element {
  const { visible, children } = props;
  return (
    <Transition visible={visible} cssPrefix="fade">
      <div>{children}</div>
    </Transition>
  );
}

export function FadeSlide(props: Omit<TransitionProps, 'cssPrefix'>): JSX.Element {
  const { visible, children } = props;
  return (
    <Transition visible={visible} cssPrefix="fade-slide">
      <div>{children}</div>
    </Transition>
  );
}
