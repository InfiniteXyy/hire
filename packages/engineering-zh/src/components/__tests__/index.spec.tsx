import React from 'react';
import { render, screen } from '@testing-library/react';
import EmptyView from '../empty';
import Loading from '../loading';
import { FadeSlide } from '../transition';

describe('test AutoComplete external utils', () => {
  it('should emptyView return oops ', () => {
    render(<EmptyView active={true} />);
    expect(screen.queryByText(/oops/i)).toBeTruthy();
  });

  it('should loading match snapshot', () => {
    const wrapper = render(<Loading active={true} />);
    expect(wrapper.baseElement).toMatchSnapshot('loading');
  });

  it('should transition works', () => {
    // TODO: more detailed test
    jest.useFakeTimers();
    render(
      <FadeSlide visible={true}>
        <div>123</div>
      </FadeSlide>
    );
    expect(screen.queryByText('123')).not.toBeNull();
  });
});
