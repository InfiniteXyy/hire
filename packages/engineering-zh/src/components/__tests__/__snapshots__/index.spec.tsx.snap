// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AutoComplete external utils should loading match snapshot: loading 1`] = `
<body>
  <div>
    <div
      class="loading-spinner"
    >
      <div />
    </div>
  </div>
</body>
`;
